nntrainer/dataset/iteration_queue.cpp

```diff
@@ -194,10 +194,12 @@ IterationQueue::MarkableIteration::operator=(MarkableIteration &&rhs) {
 }
 
 void IterationQueue::MarkableIteration::markSampleFilled() {
-  std::scoped_lock notify_lock_guard(notify_mutex);
+  std::unique_lock notify_lock_guard(notify_mutex);
   num_observed++;
   if (num_observed == iteration.batch()) {
+    notify_lock_guard.unlock();
```

Comment:
Why unlock here? It seems that ```num_observed``` may experience synchronization error.

Reply:
If markFilled() cannot affect num_observed or iteration.batch() and num_observed cannot affect markFilled(), you can do ```num_observed = 0;``` before unlock.

Reply:
And, I see another potential synchronization error related to ```num_observed```: you don't project ```num_observed``` of 

https://github.com/nnstreamer/nntrainer/blob/62b6c02718690b1981a53762059568b27c582dd6/nntrainer/dataset/iteration_queue.cpp#L217-L228

Are you sure that there won't be any simultaneous executions of these lines and markSampleFilled() ? (Can we sure that ```markSampleFilled()``` is never executed while another thread is running these lines?)

Reply:
In conclusion, you may need to rethink how to protect the integrity of Sample Update (setEndSample, markSampleFilled). As long as there is possibilty of simutaneous (multi-threaded) updates on a data entry, you need to think more carefully.

Reply:
makes sense. how about adding a cv?

```
std::unique_lock notify_lock_guard(notify_mutex);
notify_emptied_num_obs_cv.wait(
  notify_mutex, [this] { return num_observed != iteration.batch(); });
num_observed++;
if (num_observed == iteration.batch()) {
  notify_lock_guard.unlock();
  iq->markFilled(this);
  notify_lock_guard.lock();
  num_observed = 0;
  notify_num_emptied_cv.notify_all();
}
```

Reply:
It's over-engineering. You don't need that much complexity.

Reply:
gotcha. Thank you for your guidance!